src/histogram/histogram.cpp

```diff
@@ -15,12 +15,82 @@
 #include <boost/histogram/axis/ostream.hpp>
 #include <boost/histogram/ostream.hpp>
 
+#include <boost/mp11.hpp>
+
 #include <cassert>
 #include <vector>
 #include <sstream>
+#include <tuple>
+#include <cmath>
 
 namespace bh = boost::histogram;
 
+template <class Histogram>
+struct fill_helper {
+  fill_helper(Histogram& h, py::args args) : hist(h) {
+    auto dim = args.size();
+    if (dim == 0)
+      throw std::invalid_argument("at least one argument required");
+
+    arrs.reserve(dim);
+    for (std::size_t i = 0; i < dim; ++i) {
+      auto a = py::cast<py::array_t<double>>(args[i]);
+      if (a.ndim() > 1)
+        throw std::invalid_argument("array dim must be 0 or 1");
+      arrs.emplace_back(a, a.data());
+    }
+
+    size = arrs[0].first.size();
+    for (std::size_t i = 1; i < dim; ++i)
+      if (arrs[i].first.size() != size)
+        throw std::invalid_argument("arrays must have same length");
+    if (size == 0) // handle scalars by setting size to 1
+      ++size;
+  }
+
+  // keep function small to minimize code bloat, it is instantiated 32 times :(
+  // TODO: solve this more efficiently on the lower Boost::Histogram level
```

Comment:
By the way, the compile is noticeably slower now and the output .so is 10 MB (not sure how large it was before).